[DSIP-90][Task] Add a gRPC task plugin

npofsi opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler, we are happy that you want to help us improve DolphinScheduler! -->

## Purpose of the pull request

This pull request adds an implementation for gRPC plugin which is introduced in [DSIP-90](github.com/apache/dolphinscheduler/issues/17290). It provides an JSON api to transfer protobuf definitions between webpage and server.

Close #17290
## Brief change log
BrowserSide:
- Add gRPC plugin component in workflow space
- Add gRPC plugin UI to accept Protobuf and messages
- Translate definitions from Protobuf to JSON use protobuf.js

<img width="375" height="252" alt="image" src="https://github.com/user-attachments/assets/719c2b0c-3ff9-41e6-ab61-2f311fec657a" />
<img width="2880" height="1530" alt="image" src="https://github.com/user-attachments/assets/f82f6efd-c4b3-49d2-b331-7dbd91ce20f8" />
<img width="2880" height="1530" alt="image" src="https://github.com/user-attachments/assets/33ce3e05-2a34-4139-85ed-4639cd11b0ba" />

ServerSide:
- Written a JSON Parser to reconstruct the Protobuf from browser side
- Making gRPC call from server and check the response

## Verify this pull request

<!--*(Please pick either of the following options)*-->

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

<!--*(example:)*
- *Added dolphinscheduler-dao tests for end-to-end.*
- *Added CronUtilsTest to verify the change.*
- *Manually verified the change by testing locally.* -->

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contains incompatible change, you should also add it to `docs/docs/en/guide/upgrade/incompatible.md`

npofsi committed:
[Chore] Prepare pom.xml for grpc plugin

npofsi committed:
[DSIP][Chore] Add basic files for grpc plugin

npofsi committed:
[Chore] Clean line breaks

npofsi committed:
[Feature] Add ui basics for gRPC plugin

npofsi committed:
[Feature] Add ui basics for gRPC plugin

npofsi committed:
[Feature] Add Parameters

npofsi committed:
[Feature] Add Test Files

npofsi committed:
[Chore] Add Channel Factory

npofsi committed:
[Chore] Register plugin at frontend

npofsi committed:
[Chore] Add Code License

npofsi committed:
[Feature] Register plugin icon

npofsi committed:
[Feature] Add url field in plugin

npofsi committed:
[Feature] Add fields

npofsi committed:
[Feature] Add protobuf parser deps

npofsi committed:
[Feature] Add Locales

npofsi committed:
[Feature] Replace texts

npofsi committed:
[Chore] Typo fix

npofsi committed:
[Feature] Typo fix

npofsi committed:
Merge branch 'subr' of https://github.com/npofsi/dolphinscheduler into subr

npofsi committed:
[Chore] Clean unused properies

npofsi committed:
[Feature] match variables

npofsi committed:
[Chore] Add initial value

npofsi committed:
[Chore] Change method hint

npofsi committed:
[Fix] Use gRPC Status Code desc

npofsi committed:
[Feature] Protobuf post format

npofsi committed:
[Feature] Add ProtobufJS Parser

npofsi committed:
[Feature] Add task handle

npofsi committed:
[Test] Add Basic Test Assets

npofsi committed:
[Fix] Hint correction

npofsi committed:
[Test] Add Test Generator

npofsi committed:
Test: add test server

npofsi committed:
Feat: add service definition validator

npofsi committed:
Feat: add JSON definition parser

npofsi committed:
[Chore]Change protobuf for test options

npofsi committed:
[Test] Add test files

npofsi committed:
[Chore] Clear Imports

npofsi committed:
[Test] Add task tester

npofsi committed:
[Test] Add basic task tests for gRPC plugin

npofsi committed:
[Test] Add test for default output

npofsi committed:
[Test] Add parameters tests

npofsi committed:
[Feat] Add grpc url validator

npofsi committed:
[Chore] Clear comment

npofsi committed:
[Chore] Add license header

npofsi committed:
[Chore] Add license header for proto file

npofsi committed:
[Build] Update pnpm lock file

npofsi committed:
[Chore] Add license to lock file

npofsi committed:
[Test] Bump to JUnit5

npofsi committed:
[Test] Use JUnit5 Extension

npofsi committed:
[Test] Fix import

npofsi committed:
[Feat] Add timeout support

npofsi committed:
[Test] Fix parameters for unit test

npofsi committed:
[Test] Fix parameters test

npofsi committed:
[Test] Run spotless

npofsi committed:
[Clean] Remove unused code

npofsi committed:
[Feat] Add mapField support

npofsi committed:
[Clean] Remove unused code

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
[Test] Add parser tests

npofsi committed:
[Chore] typo fix

npofsi committed:
[Fix] Map field parser

npofsi committed:
[Test] Add tests for basic types and map fields

npofsi committed:
[Chore] run spotless

npofsi committed:
[Feat] Handle mismatched status code

npofsi committed:
[Feat] Add field label support

npofsi committed:
[Feat] Full support for map

npofsi committed:
[Test] Add oneof type

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
Merge branch 'subr' of https://github.com/npofsi/dolphinscheduler into subr

npofsi committed:
[Feat] Add oneof type and optional label support

npofsi committed:
[Fix] Synthetic oneofs must come last

Gallardot committed:
Merge branch 'dev' into subr

npofsi commented:
## PR Stage Summary (Changes or Adds)

### Frontend
- I18n Strings for field hints
- Task Icon
- Add fields in task params for gRPC Task(message, serviceDefinition, serviceDefinitionJSON, methodName, ...)
- Task UI with correction checker(especially serviceDefinition and messageJson)
- Dependency of Protobufjs

### Backend
- gRPC Task and its channel & channelFactory
- CheckCondition implementation based on gRPC Status Code
- gRPC dynamic service for runtime gRPC stub generation
- A full java implemented ProtobufJS Description Parser
- Parser implementation of basic protobuf types(namespace, message, field, nested message, primitive types, service, method(rpc), enum)
- Special field and its labels(map field, optional field, repeated field, oneof field)

### Test
- Backend task function test
- Backend task parameter test
- Workflow context binding test
- Parser test with different protobuf types
- JUnit5 gRPC test cleanup extension
- Protobuf Definitions for test(taskTester.proto, parserTester.proto) and corresponding json files in resource

@Gallardot

SbloodyS reviewed (commented):
Please add `api-test` or `e2e-test` to ensure stability. @npofsi

npofsi committed:
[Test] Add api test for gRPC task

npofsi committed:
[Chore] Comments edits

npofsi committed:
[Test] Move task api test into executor test

npofsi committed:
[Fix] Typo fix

npofsi committed:
[Dep] Bump protoc version to 3.17.3 for osx-aarch64 support

npofsi committed:
[Fix] Lint fix

npofsi committed:
[Fix] Lint fix

npofsi committed:
[Chore] Disable helper link

npofsi committed:
[Chore] Add known dependency

npofsi committed:
[Test] Add basic files for e2e test

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
Merge branch 'subr' of https://github.com/npofsi/dolphinscheduler into subr

npofsi committed:
[UI] Add class name

npofsi committed:
[Test] Add support for multiple editors

npofsi committed:
[Chore] Add gRPC to plugins config

npofsi committed:
[Test] Add e2e test for gRPC task

npofsi committed:
[Test] Use official image of grpc

npofsi committed:
[Chore] Run spotless

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
[Test] Add e2e test for success case

npofsi committed:
[Test] Add await for tab switches

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
[Test] Add e2e test for failed  case

npofsi committed:
Merge branch 'subr' of https://github.com/npofsi/dolphinscheduler into subr

npofsi committed:
[Test] Fix curl http 0.9 error

npofsi committed:
[Test] Wait for dialog dismiss

npofsi committed:
[Test] Add delay tester for modal mask

npofsi committed:
[Test] Run Spotless

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
[Test] Click save button use script

npofsi committed:
[Test] Get web driver

npofsi committed:
[Test] Use script click

npofsi committed:
[Test] Test wait

npofsi committed:
[Test] Fixed window size

npofsi committed:
[Fix] Use explicit channel for socket address

npofsi committed:
[Test] Add await for tab change

npofsi committed:
[Test] Use click method

npofsi committed:
[Test] Add wait time

npofsi committed:
[Test] Use full exception stack

SbloodyS commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/java/org/apache/dolphinscheduler/api/test/cases/ExecutorAPITest.java:
```diff
@@ -145,4 +145,53 @@ public void testStartWorkflowInstance() {
         }
     }
 
+    @Test
+    @Order(2)
+    public void testGrpcWorkflowInstance() {
```
We should move this test case into `GrpcTaskAPITest` since this is an new task type.

npofsi replied:
This will be moved at following commits.

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/workflow/BaseWorkflowE2ETest.java:
```diff
@@ -167,8 +202,15 @@ protected TaskInstanceTab.Row untilTaskInstanceFailed(String workflowName, Strin
+                    TaskInstanceTab taskInstanceTab = projectPage
+                            .goToTab(TaskInstanceTab.class);
+
+                    List<TaskInstanceTab.Row> taskInstances = taskInstanceTab
```
Why change this? The current detection process is normal and does not need to be modified.

npofsi replied:
The new awaits added to E2E are ensuring the behaviour stability. Maybe I need some help here.

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/pages/LoginPage.java:
```diff
@@ -70,7 +75,13 @@ public NavBarPage login(String username, String password) {
         buttonLogin().click();
 
-        WebDriverWaitFactory.createWebDriverWait(driver).until(ExpectedConditions.urlContains("/home"));
+        await().ignoreException(NoSuchElementException.class)
+                .untilAsserted(() -> assertThat(driver)
+                        .as("can not go to menu")
+                        .matches(it -> it
+                                .findElement(By.xpath(
+                                        "//div[contains(@class, 'tab-horizontal')]//div[contains(@role,'menubar')]"))
+                                .isDisplayed()));
```
Same here. Why change this?

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/pages/project/ProjectDetailPage.java:
```diff
@@ -53,12 +58,24 @@ public <T extends Tab> T goToTab(Class<T> tab) {
             menuWorkflowDefinition().click();
             WebDriverWaitFactory.createWebDriverWait(driver)
                     .until(ExpectedConditions.urlContains("/workflow-definition"));
+
+            await().ignoreException(NoSuchElementException.class)
+                    .untilAsserted(() -> assertThat(driver)
+                            .as("workflow definition page not found")
+                            .matches(it -> it.findElement(By.className("n-data-table-base-table-body")).isDisplayed()));
+
             return tab.cast(new WorkflowDefinitionTab(driver));
         }
         if (tab == WorkflowInstanceTab.class) {
             menuWorkflowInstances().click();
             WebDriverWaitFactory.createWebDriverWait(driver)
                     .until(ExpectedConditions.urlContains("/workflow/instances"));
+
+            await().ignoreException(NoSuchElementException.class)
+                    .untilAsserted(() -> assertThat(driver)
+                            .as("workflow instance page not found")
+                            .matches(it -> it.findElement(By.className("n-data-table-base-table-body")).isDisplayed()));
+
```
Same here. Why change this?

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/pages/project/ProjectPage.java:
```diff
@@ -132,13 +140,21 @@ public ProjectPage verifyAssignedWorkerGroup(String project, String workerGroup)
+        await().untilAsserted(() -> assertThat(driver)
+                .as("not in project detail page")
+                .matches(it -> it.findElement(By.className("n-menu-item-content")).isDisplayed()));
+
```
Same here. Why change this?

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/pages/project/workflow/WorkflowDefinitionTab.java:
```diff
@@ -72,6 +73,11 @@ public WorkflowDefinitionTab(RemoteWebDriver driver) {
     public WorkflowForm createWorkflow() {
         buttonCreateWorkflow().click();
 
+        await().ignoreException(NoSuchElementException.class)
+                .untilAsserted(() -> assertThat(driver)
+                        .as("created workflow must be displayed")
+                        .matches(it -> it.findElement(By.className("n-collapse-item")).isDisplayed()));
+
```
Same here. Why change this?

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/resources/docker/grpc-task/docker-compose.yaml:
```diff
@@ -0,0 +1,53 @@
+      retries: 120
+  greeterServer:
+    image: grpc/java-example-hostname:1.75.0
+#    environment:
```
```suggestion
```

Please remove this.

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-core/src/main/java/org/apache/dolphinscheduler/e2e/core/DolphinSchedulerExtension.java:
```diff
@@ -106,8 +107,9 @@ public void beforeAll(ExtensionContext context) throws IOException {
         driver.manage().timeouts()
                 .implicitlyWait(Duration.ofSeconds(1))
                 .pageLoadTimeout(Duration.ofSeconds(5));
+        Dimension size = new Dimension(1920, 1080);
```
Why change this?

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/pom.xml:
```diff
@@ -0,0 +1,297 @@
+        <dependency>
+            <groupId>com.google.j2objc</groupId>
+            <artifactId>j2objc-annotations</artifactId>
+            <version>3.0.0</version>
```
Dependeny version should be setting in `bom/pom.xml`.

npofsi replied:
Get it.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/pom.xml:
```diff
@@ -0,0 +1,297 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.13.2</version>
```
Same here.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/pom.xml:
```diff
@@ -0,0 +1,297 @@
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <version>3.5.0</version>
+                <executions>
+                    <execution>
+                        <id>enforce</id>
+                        <goals>
+                            <goal>enforce</goal>
+                        </goals>
+                        <configuration>
+                            <rules>
+                                <requireUpperBoundDeps />
+                            </rules>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+        <extensions>
+            <extension>
+                <groupId>kr.motd.maven</groupId>
+                <artifactId>os-maven-plugin</artifactId>
+                <version>1.7.1</version>
+            </extension>
+        </extensions>
```
Why adding this?

I think we dont't need it.

npofsi replied:
The test depends on a protobuf compiler, protoc, which is a executable maven plugin using os type. The os-maven-plugin is the plugin to get system os type and pass the type to maven to get the specific version of protoc.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/protobufjs/JSONDescriptorParser.java:
```diff
@@ -0,0 +1,297 @@
+                break;
+            }
+        }
+        return String.join(".", packageNameNS);
```
Please using `Constants` instead of hard code it.

npofsi replied:
Get it.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/test/java/org/apache/dolphinscheduler/plugin/task/grpc/.gitignore:
We should remove unnessnary file.

npofsi replied:
The tests use protoc to compile the .proto file to java classes, and this ignore file is used to unselect the generated files, because these files will generate at test-compose stage.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/test/java/org/apache/dolphinscheduler/plugin/task/grpc/GrpcCleanupExtension.java:
```diff
@@ -0,0 +1,176 @@
+
+    private final List<CleanupTarget> cleanupTargets = new ArrayList<>();
+
+    private static final Logger logger = LoggerFactory.getLogger(GrpcCleanupExtension.class);
```
```suggestion
```

Using `@Slf4j` instead.

npofsi replied:
Get it.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/test/resources/taskTester.json:
```diff
@@ -0,0 +1,61 @@
+      }
+    }
+  }
+}
```
Please leave an blank line at the end of each file.

npofsi replied:
Get it.

SbloodyS commented on dolphinscheduler-ui/pnpm-lock.yaml:
Please revert unnessnary change in this file.

npofsi replied:
Get it.

SbloodyS commented on tools/dependencies/known-dependencies.txt:
```diff
@@ -300,6 +300,7 @@ perfmark-api-0.23.0.jar
 proto-google-common-protos-2.0.1.jar
 protobuf-java-3.17.2.jar
 protobuf-java-util-3.17.2.jar
+protobuf-java-util-3.17.3.jar
```
Why do we need to introduce a new version? Since we already have `protobuf-java-util-3.17.2`.

npofsi replied:
The 3.17.3 is the first protobuf version with a osx-supported protoc. If we have no need with test on osx, this can be reverted.

npofsi committed:
[Test] Remove comment

Co-authored-by: xiangzihao <zihaoxiang@apache.org>

npofsi committed:
[Test] Revert unused e2e changes

npofsi committed:
[Test] Revert changes on BaseWorkflow

npofsi committed:
[Chore] Add empty lines

npofsi committed:
[Chore] Clean unused packages

npofsi committed:
[Build] Use bom to manage deps

npofsi committed:
[UI] Revert npm lock file

npofsi committed:
[Chore] Move constants to Constants class

npofsi committed:
[Test] Use slf4j for unit test

npofsi committed:
[Test] Move grpc api test as independent test

npofsi committed:
[Test] Remove order

npofsi committed:
[Test] Clean import

npofsi committed:
[Build] Update lock file

npofsi committed:
[Build] Update lock file

npofsi committed:
[Build] Revert bump protobuf version

npofsi committed:
[Build] Revert bump protobuf java util version

npofsi committed:
[Test] Fix import

SbloodyS committed:
Merge branch 'dev' into subr

npofsi committed:
[Test] Fix message input

npofsi committed:
[Test] Run spotless

npofsi committed:
[Test] Relocate editor lines before writing message

npofsi committed:
[Test] Ensure lines in editor

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/protobufjs/JSONDescriptorHelper.java:
```diff
@@ -0,0 +1,41 @@
+    public static Root ProtobufFromJSON(String json) throws JsonProcessingException {
+        ObjectMapper mapper = JsonMapper.builder()
+                .build();
+        return mapper.readValue(json, Root.class);
```
Using `JsonUtils` instead.

npofsi committed:
[Test] Use get instead array index

npofsi committed:
[Test] Fix typo

npofsi committed:
[Test] Fix cursor offset

npofsi committed:
[Test] Remove unnecessary wait

npofsi committed:
[Chore] Use JSONUtil

npofsi committed:
[Test] Align editor to top

npofsi committed:
[Test] Trace editors before content

npofsi committed:
[Test] Trim content

npofsi committed:
[Test] Run spotless

npofsi committed:
[Test] Locate editors at run time

npofsi committed:
[Test] Disable side panel pinning

npofsi committed:
[Test] Fix method name

npofsi committed:
[Chore] Bump monaco editor version to 0.50.0

npofsi committed:
[Test] Remove unnecessary flag

npofsi committed:
Revert "[Chore] Bump monaco editor version to 0.50.0"

This reverts commit 7486230bfc40303829b336ea24503a9518130aef.

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/pom.xml:
```diff
@@ -0,0 +1,174 @@
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
```
```suggestion
       
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/pom.xml:
```diff
@@ -0,0 +1,174 @@
+            <artifactId>dolphinscheduler-task-api</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.dolphinscheduler</groupId>
+            <artifactId>dolphinscheduler-common</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-services</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-netty</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-netty-shaded</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-protobuf</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-stub</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.google.protobuf</groupId>
+            <artifactId>protobuf-java-util</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
Why add these, these is not needed? if we only need this in test, we should change the scope to test, and the version shouldn't be managed at bom if we only need this in UT

npofsi replied:
These are used at the construction stage for grpc client. 
grpc-service & grpc-netty: create managed channel to an remote address.
grpc-protobuf: convert from primitive protobuf descriptors to protobufs in grpc domain.
protobuf-java-util: convert json descriptors to structed protobuf descriptors.
grpc-stub: create a client accessing to corresponding managed channel.

ruanwenjun replied:
If so, it's better to remove the provided scope.

npofsi replied:
get

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/GrpcParameters.java:
```diff
@@ -0,0 +1,57 @@
+     * Connect Timeout
+     * Unit: ms
+     */
+    private int connectTimeout = 0;
```
```suggestion
    private long connectTimeoutMs = 0L;
```

npofsi committed:
[Chore] Bump monaco editor to 0.53.0

npofsi committed:
[Chore] Fix deps for monaco editor

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/protobufjs/JSONDescriptorParser.java:
```diff
@@ -0,0 +1,297 @@
+import com.google.protobuf.DescriptorProtos;
+import com.google.protobuf.Descriptors;
+
+public class JSONDescriptorParser {
```
Is this needed? If we directly receive the proto content rather than json schema, then we don't need this? This seems meaningless.

npofsi replied:
To get a structed protobuf descriptor, the only choice is `protoc` or `protobufjs`. The first one needs the dependency of the binary executable protoc in paths. To avoid an additional dependency, we choose protobufjs. We use it to generate a structed protobuf descriptor representated in json at frontend, and reconstruct the protobuf descriptor at backend. We can not use a .proto text as a structed definition directly (which is different from json).

ruanwenjun replied:
We can generate the `.proto` file, then avoid doing the complex parsing here.

npofsi replied:
We can't use `.proto` file directly, because the ` .proto` file is designed as a static definition, not a format to transfer data. The inner transmission type of protobuf is a binary format. The file needs to be compiled to a builder of binaries (by `protoc` or `protobufjs`). So that we can use the compiled dist to create a runtime definition (or protobuf descriptor) dynamically. And we choose to compile it at front end.

npofsi replied:
All of we need at runtime construction are the things after the protobuf compile. The protobuf workflow is demostrated in following image. 
https://i0.wp.com/lab.wallarm.com/wp-content/uploads/2023/12/253-min.jpg?w=770&ssl=1

npofsi replied:
Of course we can save the `.proto` into backend and compile it at runtime. But there are some trade offs.
For using `protobufjs`, running a js engine to interprete the `protobufjs` in java is not an efficient and clear solution.
For using `protoc`, we always need additional dependencies, even `libprotoc` or `libprotoc-java` needs a binary executable `protoc`.

npofsi replied:
In the current solution, we compile the .proto at frontend, so we can validate user input at realtime. And the json descriptor of the format between frontend and backend is defined by `protobufjs`, which is more reliable then running a js engine at backend, the only problem is an implementation of reconstructor.

ruanwenjun replied:
Not all requests originate from the UI; they may also come from the API, so the frontend validation isn't particularly important, we still need to perform all validation on the backend.

I ask gpt, it tells me that we can use some lib to parse the proto, we don't need to write the parser here.

npofsi replied:
Certainly, we also do parse stage check for input.

If a java lib can compile a .proto at run time, please provide to me. It must be useful.

npofsi replied:
If the DynamicMessage had mentioned by chatgpt, it has already used in current solution.

npofsi committed:
[Chore] Use united variable name

npofsi committed:
[Test] Remove unused pom

npofsi committed:
[Test] Run Spotless

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/protobufjs/JSONDescriptorParser.java:
```diff
@@ -0,0 +1,297 @@
+                        fileDescriptorProtoBuilder.addService(parseService(name, (Service) ns));
+                    } else {
+
+                    }
```
Why adding empty else here?

npofsi replied:
Unnecessary branches, will be removed.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/protobufjs/JSONDescriptorParser.java:
```diff
@@ -0,0 +1,297 @@
+
+                } else if (pbObject instanceof Method) {
+
+                }
```
Why adding empty else here?

npofsi replied:
Also

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/protobufjs/JSONDescriptorParser.java:
```diff
@@ -0,0 +1,297 @@
+                    descriptorProtoBuilder.getOneofDeclList();
+            oneofDescriptorProtos.forEach((oneof) -> {
+
+            });
```
Why adding empty logic here?

npofsi replied:
Also

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/pages/project/workflow/task/GrpcTaskForm.java:
```diff
@@ -0,0 +1,75 @@
+
+            @FindBy(className = "input-method-name"),
+            @FindBy(tagName = "input")
+    })
```
```suggestion
    @FindBys({
            @FindBy(className = "input-method-name"),
            @FindBy(tagName = "input")
    })
```

npofsi committed:
Revert "[Chore] Bump monaco editor to 0.53.0"

This reverts commit cc79d8a3f373379556018718b76fc687acc85b66.

npofsi committed:
[Test] Remove gRPC e2e test

npofsi committed:
[Chore] Clean pom

npofsi committed:
[Chore] Remove unnecessary branches

npofsi committed:
[Test] Add api test files

npofsi committed:
Revert "[Chore] Clean pom"

This reverts commit beb4203cbeceb21f885ca0a11e7aa6677365a327.

npofsi committed:
[Test] Fix docker-compose.yaml

npofsi committed:
[Chore] Clean pom

npofsi committed:
[Test] Add gRPC task api test

Gallardot committed:
Merge branch 'dev' into subr

Gallardot committed:
Merge branch 'dev' into subr

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
[Feat] Add client default ssl/tls support

Gallardot committed:
Merge branch 'dev' into subr

npofsi committed:
chore: add proto helper file

npofsi committed:
[Feat] Add parameter format check

npofsi committed:
[Clean] clean dependencies

SbloodyS committed:
Merge branch 'dev' into subr

npofsi committed:
[Fix] Dependency version of jsr305

npofsi committed:
Merge branch 'subr' of https://github.com/npofsi/dolphinscheduler into subr

npofsi committed:
[Docs] Add zh docs for gRPC task

npofsi committed:
[Docs] Add en docs of gRPC task

npofsi committed:
[Docs] Run spotless apply

npofsi committed:
[Docs] Typo correction

npofsi committed:
Merge branch 'apache:dev' into subr

Gallardot committed:
Merge branch 'dev' into subr

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/test/java/org/apache/dolphinscheduler/plugin/task/grpc/.gitignore:
Please remove unnessnary file.

npofsi committed:
[Test] Remove .gitignore for generated classes

ruanwenjun commented on dolphinscheduler-ui/package.json:
```diff
@@ -24,6 +24,7 @@
     "nprogress": "^0.2.0",
     "pinia": "^2.0.22",
     "pinia-plugin-persistedstate": "^2.2.0",
+    "protobufjs": "^7.5.4",
```
Please add license https://dolphinscheduler.apache.org/en-us/docs/3.3.1/contribute/join/DS-License

SbloodyS committed:
Merge branch 'dev' into subr

npofsi committed:
[Docs] Add denpendency license

npofsi committed:
[Chore] Clean code

SbloodyS committed:
Merge branch 'dev' into subr

ruanwenjun commented on dolphinscheduler-bom/pom.xml:
```diff
@@ -958,6 +1036,15 @@
+
+        </repository>
+    </repositories>
+
```
Please remove this

ruanwenjun commented:
Please fix the new issue
<img width="1664" height="628" alt="image" src="https://github.com/user-attachments/assets/5936ce03-cf13-429f-b70c-94840426d323" />

npofsi committed:
[POM] Remove unused repository

npofsi committed:
Merge branch 'subr' of https://github.com/npofsi/dolphinscheduler into subr

npofsi committed:
[Chore] Change name of inner package

npofsi committed:
[Refactor] Reconstruct parser for recognizability

Move methods from parser class into seperated classes in types package for recognizability. Control flows are not changed

npofsi committed:
[Test] Change enum value name in tester proto

npofsi committed:
[Chore] Remove unused constructors

npofsi committed:
[Chore] Remove line

npofsi committed:
[Test] Add license headers

npofsi committed:
[Chore] Cleanup code

npofsi committed:
[Chore] Remove lines

npofsi committed:
[Chore] Remove unused parameters

npofsi committed:
[Chore] Clean generated class

npofsi committed:
[Chore] Use semantic exceptions

npofsi committed:
[Chore] Use package specific parseInt

npofsi committed:
[Chore] Clean code

npofsi committed:
[Chore] Clean code

npofsi committed:
[Chore] Hide implicit public constructors

npofsi committed:
[Chore] Add access control for class properties

npofsi committed:
[Chore] Clean code

npofsi committed:
[Chore] Move generated code dir

npofsi committed:
[Clean] Clean unused code

npofsi committed:
[Feat] Add request result to log

npofsi committed:
[Chore] Remove err receiver

Gallardot reviewed (dismissed):
LGTM

ruanwenjun commented on docs/docs/en/guide/task/grpc.md:
```diff
@@ -0,0 +1,57 @@
+| gRPC Url                 | gRPC request URL, must use the `hostname:port` format                                                                                                           |
+| gRPC Credential Type     | Supports None (insecure), and client default SSL/TLS credential types                                                                                           |
+| Protobuf Definition      | Protobuf code defining the service in the `.proto` file, will be converted to JSON Descriptor when saved                                                        |
+| Protobuf Definition JSON | Used for creating tasks via API, generated from the above definition as [JSON Descriptor](https://github.com/protobufjs/protobuf.js/#using-json-descriptors)    |
```
Is this need to appear at docs? We don't have a task parameter named `Protobuf Definition JSON`in the task form.

ruanwenjun commented on .github/workflows/e2e.yml:
```diff
@@ -192,6 +192,7 @@ jobs:
       - name: Run Test
         run: |
           ./mvnw -B -f dolphinscheduler-e2e/pom.xml -am \
+            -DtrimStackTrace=false \
```
```suggestion
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-grpc/src/main/java/org/apache/dolphinscheduler/plugin/task/grpc/GrpcParameters.java:
```diff
@@ -0,0 +1,67 @@
+
+    private String condition;
+
+    @JsonProperty("connectTimeout")
```
```suggestion
    @JsonProperty("connectTimeoutMs")
```

ruanwenjun reviewed (dismissed):
LGTM

npofsi committed:
[CI] Remove unused flag

npofsi committed:
[Doc] Remove invisible parameter

npofsi committed:
Merge branch 'dev' into subr

npofsi committed:
[API] Use Semantic Timeout Property for gRPC Task

SbloodyS committed:
Merge branch 'dev' into subr

npofsi committed:
[Test] Test definition correction

npofsi committed:
[Test] Test definition correction

ruanwenjun reviewed (approved):
LGTM
